Feature/troth enhancements v2

troth opened the pull request:
This pull request contains a bunch of bug fixes, enhancements and new features that I have made over the past few years. I have merged my changes onto the current master.

tkrajina commented:
Hi,

First of all sorry for my late reply and thank you for your pull request. It would be much easier for me if this was a series of pull requests, since most commits are atomic features. 

Before merging it there are a few **minor** issues:

1. First is the `--quiet` option, which is very similar to `--changed`. The only difference is than `--changed` outputs `Not changed`. If you have hundreds of unchanged projects, that could be an issue, but in that case you can use `-g` (group by output) and it will be shown only once. Isn't this good enough to you?

2. `.repo/manifest.xml` is basically a list of projects you want to be used with `git multi`? I don't have anything against it but the problem with accepting a feature like this is that it's not widely used. If somebody later makes a pull request on that code, I can't experiment myself (because I'm not using those `manifest.xml` files myself) and I have to blindly maintain a piece of code in the. Maybe it would be easier to create a simple script which would create `.gitmulti` from your `manifest.xml` file?

3. Also (if we agree on both files), `.multigit_ignore` and `.gitmulti` should both be handled in the same place, I like your `get_project_list()` better than the place where I'm currently handling `.multigit_ignore`.

4. Now there is a new config file `.multigit_ignore` and `.gitmulti` -- the names should be more consistent. Now that I think about it, your `gitmulti`makes more sense, but since `.multigit_ignore` is older, I suggest `.multigit` for the new file). Also, I'm not completely sure if I want to have both of those files, in a way, having one is equivalent to having the other, and creates some grey areas when a user has both and a project both explicitly listed in projects and ignored. What then?

Maybe (like I suggested with `manifest.xml`) just create a script which will "transform" your `.gitmulti` into `.multigit_ignore`? Maybe we can add a new directory `helper_scripts` in the project for scripts like this one.

This list may seem long, but for your 12 commits I only have 4 issues, that means I'm pretty happy with most of your pull request :)

If you want to further discuss points 1-4, I'm here. Otherwise, I'll merge your PR and revert 3 comits (for `--quiet`, `manifest.xml` and `.gitmulti`). Let me know your thoughts.

tk

troth commented:
Thanks for looking over my changes.

1. Myself and a few other people have found the '-q' option to be quite useful since it eliminates all repos that are unchanged and lets you focus on what has actually changed without having to parse what, to us, is not useful information.

2. Google repo manifest files are in use in many places (the largest being Android development). I have worked multiple jobs in the last few years that have used these manifest files. It is extremely convenient to have git multi just automagically find them and use them. Having to run another tool to convert the manifest into another format (that is less widely used and of unknown robustness) is less than optimal. For example, every time you would run 'repo sync', you would have to run the converter in case the repo sync pulled in more git repos via updating the manifest. The manifest file format is pretty simple XML, has been around for a while now and has not seem much churn if any in a while. Not supporting manifest natively makes 'git-plus' a little less useful to a lot of people. The maintenance you fear should be minimal if it even exists. Plus, it's not a big deal for me to help maintain this feature.

3. The '.git*' config file name also makes for sense to me since then you see all our .git config files lumped together when you do an 'ls'. In the end, I don't really care what the name of the file is. The real problem that I'm trying to solve is the case where you have a bunch of repos, but they are not all placed under the same parent (e.g. you have one parent project with many child sub projects). Without my changes, 'git-multi' can not handle that without walking all subdirs looking for .git dirs (and this method is slow and prone to error -- consider an openembedded/yocto build which can have many git repos that get cloned by the bitbake system but which you do not want git-multi to try to walk; I had implemented this, but declined to submit a patch for it since the performance was poor at best). Google repo manifests seem to solve this the best for me, but my .gitmulti config file also addresses this issue as alternative if repo manifests are not available. Also consider that a user might not have to maintain the repo manifests since they may be maintained by the upstream project or one person on a team, whereas each user of the git-multi will most likely have to maintain their own .gitmulti file). As you can see, I am quite biased in favor of using repo manifest files if they are available.

4. This sounds like a future enhancement. Should not be too difficult. I could whip up a patch for this. I also have an pending patch that allows running git-multi from any subdir of the directory within the parent of all of the repos under consideration. If walks *up* the tree looking for .gitmulti or repo manifest files. If those are not found, it reverts to using the current directory. This is quite useful if you want to run 'git-multi' to search all repos for usage or definitions outside of the repo you are currently working on without having to cd to the top dir. Merging the .multigit_ignore and .gitmulti into a single file is probably the best approach since it opens up opportunity to do configuration stuff similar to what ~/gitconfig or <repo>/.git/config allow you to do (I'm reluctant to suggest moving git-multi config stuff into the .gitconfig files though).

troth commented on gitutils/__init__.py:
```diff
@@ -21,10 +21,10 @@ def execute_command(command, output=True, prefix='', grep=None):
-    if cmdout is None:
-        return (0, "")
+    if cmdout is None or not cmdout.strip():
+        return (p.returncode, cmdout.strip())
```
More testing revealed a bug in this line. The return should be like this:

```python
+        return (not p.returncode, cmdout.strip())
```

The results of the bug is output like this:

```
$ git multi -g
--------------------------------------------------------------------------------
Executing git status -s
--------------------------------------------------------------------------------
Progress: ## Using project list from directory search: /home/troth/git/git-plus/foo
* * * *
--------------------------------------------------------------------------------

test-1, test-2, test-3, test-4:
	Error executing:['git', 'status', '-s']
```

When the expected output is this:

```
$ git multi -g
--------------------------------------------------------------------------------
Executing git status -s
--------------------------------------------------------------------------------
Progress: ## Using project list from directory search: /home/troth/git/git-plus/foo
* * * *
--------------------------------------------------------------------------------

test-1, test-2, test-3, test-4:
	OK
```

I will integrate a fix for this and submit a reworked patchset.

troth commented:
I'm closing this pull request since I have fixed some issues with this patch set and split out the commits you have concerns about. The new patchset is here:

- https://github.com/troth/git-plus/commits/feature/troth-enhancements-v3

I need to do a bit more testing before I submit another pull request.

I also have another branch which builds on the above (includes the features you had concerns on plus a few more changes):

- https://github.com/troth/git-plus/commits/feature/goog-repo-search-v2
